### DISCUSSION TITLE
Bitrate in Admission Webhooks

### DISCUSSION THREAD

**Discussion by irlkitcom** [OvenMediaLabs/OvenMediaEngine#1579] [Ideas] [answered]
Add the bitrate from the client in the admission webhooks to have the option to deny incoming streams that are over a certain amount of bitrate.

**Comment by naanlizard** [comment]
This would be useful for us as well

**Comment by getroot** [comment, marked as answer]
AdmissionWebhooks operate before the client sends stream data. Therefore, OME cannot calculate the bitrates at that time.
Please refer to the Alert function below, which was created for that purpose.
https://airensoft.gitbook.io/ovenmediaengine/alert

**Reply by irlkitcom** [reply]
This is perfect, I didn't see it in the documentation at first.

**Reply by bchah** [reply]
A question on this - in the scenario where we need to block high bitrates - will the Alert function notify the control server *before* OME starts processing the incoming stream data? Or does this happen concurrently? The scenario I am wondering about is if somebody tried to "flood" the server with a 250Mbps stream, will we be able to fully block the transmission before it begins?

**Reply by getroot** [reply]
Because the bitrates are the number of bits transmitted per second, it is impossible for OME to know before they are received.

If it is essential, it is recommended to limit the maximum bandwidth per TCP/UDP session using external software such as Linux's `tc`.

**Comment by naanlizard** [comment]
Doesn't ome get a biterate value from the rtmp data? I'm wondering how we
get the value in the API so exact matching the setting in the encoder.

**Reply by bchah** [reply]
I think RTMP lets you 'declare' a bitrate as part of opening the stream, but that doesn't necessarily reflect the real-world bitrate reaching the server.

**Reply by getroot** [reply]
Yes, the bitrates value that OBS sends is just numeric meta information.
All encoders strive to achieve compression close to the set bitrates, but in fact it is a very technically difficult problem. You can achieve the target bitrates almost exactly by using two-pass encoding, but it will be difficult to do it live rather than in a file.
This is especially true because VBR does not have the goal of compressing to the set bitrates.

**Comment by naanlizard** [comment]
True, it certainly doesn't reflect e.g. VBR variations in bitrate, which is
my biggest gripe. But it's usually pretty good for good faith users, who
are all probably using obs anyway which seems to do well with it

**Comment by naanlizard** [comment]
Yes, obviously VBR etc are special cases. In my experience, OBS' x264 CPU
encoder is almost always extremely accurate with its encoding bitrate usage
